syntax = "proto3";

// Define the package name for the service.
package chat;

// Define the chat service, which has two methods: JoinChat and SendMessage.
service ChatService {
  // Join the chat, returning a stream of chat messages.
  rpc JoinChat (JoinRequest) returns (stream ChatMessage) {}

  // Send a chat message to another user.
  rpc SendMessage (MessageRequest) returns (Empty) {}
}

// Define the request for the JoinChat method, which takes a username.
message JoinRequest {
  string username = 1;
}

// Define the request for the SendMessage method, which takes the sender's username,
// the recipient's username, and the content of the message.
message MessageRequest {
  string sender_username = 1;
  string recipient_username = 2;
  string content = 3;
}

// Define the response for the SendMessage method, which is an empty message.
message Empty {}

// Define the message that is sent back and forth on the chat stream.
message ChatMessage {
  string sender_username = 1;
  string content = 2;
}
